src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -1075,6 +1073,65 @@ zkutil::ZooKeeperPtr StorageReplicatedMergeTree::getZooKeeperIfTableShutDown() c
     return maybe_new_zookeeper;
 }
 
+std::vector<String> StorageReplicatedMergeTree::getZookeeperZeroCopyLockPaths() const
+{
+    const auto settings = getSettings();
+    if (!settings->allow_remote_fs_zero_copy_replication)
+    {
+        return {};
+    }
+
+    const auto & disks = getStoragePolicy()->getDisks();
+    std::set<DataSourceType> disk_types_with_zero_copy;
+    for (const auto & disk : disks)
+    {
+        if (!disk->supportZeroCopyReplication())
+            continue;
+
+        disk_types_with_zero_copy.insert(disk->getDataSourceDescription().type);
+    }
+
+    const auto actual_table_shared_id = getTableSharedID();
+
+    std::vector<String> result;
+    result.reserve(disk_types_with_zero_copy.size());
+
+    for (const auto & disk_type: disk_types_with_zero_copy)
+    {
+        auto zero_copy = fmt::format("zero_copy_{}", toString(disk_type));
+        auto zero_copy_path = fs::path(settings->remote_fs_zero_copy_zookeeper_path.toString()) / zero_copy;
+
+        result.push_back(zero_copy_path / actual_table_shared_id);
+    }
+
+    return result;
+}
+
+void StorageReplicatedMergeTree::dropZookeeperZeroCopyLockPaths(zkutil::ZooKeeperPtr zookeeper, std::vector<String> zero_copy_locks_paths,
+                                                                Poco::Logger * logger)
+{
+    for (const auto & zero_copy_locks_root : zero_copy_locks_paths)
+    {
+        auto code = zookeeper->tryRemove(zero_copy_locks_root);
+        if (code == Coordination::Error::ZNOTEMPTY)
+        {
+            LOG_WARNING(logger, "Zero copy locks are not empty for {}. There are some lost locks inside."
+                              "Removing them all.", zero_copy_locks_root);
+            zookeeper->tryRemoveRecursive(zero_copy_locks_root);
+        }
+        else if (code == Coordination::Error::ZNONODE)
+        {
+            LOG_INFO(logger, "Zero copy locks directory {} is absent on ZooKeeper.", zero_copy_locks_root);
+        }
+        else if (code != Coordination::Error::ZOK)
+        {
+            LOG_ERROR(logger, "Zero copy locks were not completely removed from ZooKeeper, "
+                              "{} still exists and may contain some garbage.", zero_copy_locks_root);
+            throw zkutil::KeeperException::fromPath(code, zero_copy_locks_root);
+        }
```

Comment:
This seems to be impossible:
https://github.com/ClickHouse/ClickHouse/blob/546484d46bc3dc01b1e88404dea7170a514871cc/src/Common/ZooKeeper/ZooKeeper.h#L271-L275

Reply:
But it makes sense to write this message if `tryRemoveRecursive` returns false

Reply:
Do you mean that the is no other reason for `code != Coordination::Error::ZOK` after I handle `ZNOTEMPTY` and `ZNONODE` in the if-branches before?

Reply:
yes (and since you don't pass a version to `tryRemove`)